osh/word_eval.py

```diff
@@ -1461,12 +1465,18 @@ def _EvalBracedVarSub(self, part, part_vals, quoted):
             elif part.prefix_op.id == Id.VSub_Bang:
                 if (part.bracket_op and
                         part.bracket_op.tag() == bracket_op_e.WholeArray):
-                    if vsub_state.has_test_op:
-                        # ${!a[@]-'default'} is a non-fatal runtime error in bash.  Here
-                        # it's fatal.
-                        op_tok = cast(suffix_op.Unary, UP_op).op
-                        e_die('Test operation not allowed with ${!array[@]}',
-                              op_tok)
+                    # undef -> empty array
+                    val = self._ProcessUndef(val, part.token, vsub_state)
+
+                    # ${!a[@]-'default'} is a non-fatal runtime error in bash.
+                    # Here it's fatal.
+                    if suffix_op_ and suffix_op_.tag() == suffix_op_e.Unary:
+                        unary_op = cast(suffix_op.Unary, suffix_op_)
+                        if consts.GetKind(unary_op.op.id) == Kind.VTest:
+                            op_tok = cast(suffix_op.Unary, UP_op).op
```

Comment:
Are op_tok and unary_op the same variable?    Can we simplify it and just use `unary_op` , and get rid of op_tok ?


I think there was some existing confusion around the suffix_op_ var and UP_op already in the code

the `UP_op` variable is intended to be used for downcasting, e.g. after `wth tagswitch()`, although it's not a hard rule

Reply:
Yeah, good point. I fixed it in 39ca6a61aef1c4a9e76ccc41b679d46ed7748a83..fc98f1a5f4243f4c5139f20b2288760680208437

Reply:
Additional fix: fc98f1a5f4243f4c5139f20b2288760680208437..ce4dc6462a3f5572021e46d2b7168202a4322721